Connection structure modification

AntonKarabaza opened the pull request:
Modify `connection.py` module to contain `Connection` interface and
`SshConnection` class implementing it.
Add `shell.py` module with `Shell` interface and `InteractiveShell` class
implementing it.
Add `channel.py` module with `Channel` interface and `SftpChannel` class
implementing it.
Remove `cmd_handler.py` module since all of his functionality was
restructured and moved to `shell.py` module.

Modify tests in `ssh_automated_tests/tests` package to use structure with
updated `connection.py` module and new `shell.py` and `channel.py` modules.
Modify `conftest.py` to create fixtures using such classes:
```
connection -> SshConnection
shell -> InteractiveShell
sftp -> SftpChannel
```
Affects #11

Commit:
Modify connection structure

Modify connection.py module to contain Connection interface and
SshConnection class implementing it.
Add shell.py module with Shell interface and InteractiveShell class
implementing it.
Add channel.py module with Channel interface and SftpChannel class
implementing it.
Remove cmd_handler.py module since all of his functionality was
restructured and moved to shell.py module.

Affects #11

Commit:
Update tests to use new connection structure

Modify tests in ssh_automated_tests/tests package to use structure with
updated connection.py module and new shell.py and channel.py modules.
Modify conftest.py to create fixtures using such classes:
	connection -> SshConnection
	shell -> InteractiveShell
	sftp -> SftpChannel

Affects #11

ivan-podorozhnyi reviewed (changes requested):
@AntonKarabaza I suggest that `Channel` abstract class must have other abstract methods besides `close()` method(for example `open_file()`) because channel is supposed to interact with files, directories etc.

AntonKarabaza commented:
@ivan-podorozhnyi Abstract method `open_file()` added to `Channel` interface as you suggested.

ivan-podorozhnyi commented:
@AntonKarabaza looks good!

extsoft commented on ssh_automated_tests/core/channel.py:
```diff
@@ -0,0 +1,36 @@
+from ssh_automated_tests.core.connection import Connection
+
+
+class Channel(ABC):
```
I would suggest renaming to `FileChannel` as this interface declares work with files. And if the advice will be applied, rename `open_file` also.

AntonKarabaza replied:
I will implement this.

extsoft commented on ssh_automated_tests/core/channel.py:
```diff
@@ -0,0 +1,36 @@
+            raise AttributeError("Given connection can't establish sftp "
+                                 "channel.")
+
+    def get_file_stat(self, file_name: str):
```
Why isn't this method declared in the interface?

extsoft commented on ssh_automated_tests/core/connection.py:
```diff
@@ -10,13 +10,17 @@ class Connection(ABC):
 
-class SshConnection:
+
+class SshConnection(Connection):
```
Looks like `open_shell` and `open_sftp` has to be separate implementations of a `Connection`.

AntonKarabaza replied:
In this realization `SshConnection` is a specific case of connection that have a possibility to open shell or sftp. Like it is in real. Is it bad to have such object?
If it's not ok, so it has to be separate implementation of `Connection` or I can just inherit `SshConnection` by `ShellConnection` and add some functionality? Because they'll have pretty similar structure except one additional method.

extsoft replied:
I'm worried that `SshConnection` has more public methods then declared in `Connection`. And for me, it means that there is a problem with the design. I just suggested my vision how to solve this.

extsoft commented on ssh_automated_tests/core/shell.py:
```diff
@@ -0,0 +1,63 @@
+        pass
+
+
+class InteractiveShell(Shell):
```
There are several methods which aren't declared in the interface `Shell`. We have to avoid that. Please redesign the interface or the class.

AntonKarabaza replied:
Why we have to declare all methods in interface? As i thought interface - it's generalized thing and our `InteractiveShell` class implements it and adds its own functionality. Here can be another implementation like `SimpleShell` which don't need such methods as `append_file_with_vim` or `send_key`.

extsoft replied:
Because interfaces allow polymorphism.

ivan-podorozhnyi replied:
I suggest this commit can be split to three commits as it affects different functionalities.
For example:
1) Changes done in connection.py
2) Adding channel.py
3) Adding shell.py and removal of cmd_handler

extsoft reviewed (changes requested):
@AntonKarabaza please add new commits instead of modifying already added.

Commit:
Modify SshConnection to implement all interface methods

Add interface SshConnection extending Connection interface.
Rename SshConnection class to BaseSshConnection and implement
SshConnection interface.

Affects #11

Commit:
Modify InteractiveShell to implement interface methods

Remove method edit_file_with_vim from InteractiveShell class.
Add methods read_output and send_key to Shell interface.

Affects #11

Commit:
Make SftpChannel implement all interface methods

Rename interface Channel to FileChannel.
Update SftpChannel to implement FileChannel.
Add method get_stat to interface FileChannel.

Affects #11

Commit:
Update tests to use restructured connection

In conftest.py use BaseSshConnection class instead of SshConnection.
In test_edit_vim.py add more test steps due to remove of
edit_file_with_vim method.
Update method renamings.

Affects #11

AntonKarabaza commented:
@extsoft Changes applied. All classes in connection structure implement interface methods. SshConnection renamed to BaseSshConnection, left not changed but get new interface SshConnection.
